### ISSUE TITLE
Will ipc-flux work with <webview>?

### ISSUE THREAD

**Issue by rkpatel33** [harryparkdotio/ipc-flux#1]
This looks really cool. I have a single browser window that has 3 webview tags in it like below.  So the processes look like this:

`Main process` <> `Render process` <> `<webview> render process`

![image](https://user-images.githubusercontent.com/3799489/29590983-6fe50b10-8751-11e7-81ef-91c05391a86f.png)

I was wondering if this will still work for direct messaging from `main process` -> `<webview> render process` since you have to pass the browser window in the following call from main:

```
ipcFlux.dispatchExternal(targetBrowserWindow, 'init');
```

Or will I have to set up a relay through the render process that spawned the <webview> processes?

Thx!

**Comment by harryparkdotio** [comment]
I haven't tested with webview, nor have I used webview before. Unfortunately I really don't know, but I'd be happy to investigate for/with you on this.

I mean, If it works with `ipc-renderer` / `ipc-main`, then it should work as expected.

If you send something to the main process from within the webview, you could reply to it from main.

**Comment by rkpatel33** [comment]
Doesn't seem to bc the messaging is a bit different between a render and a webview render process. So you can get to main from web view fine but my case requires getting to the "embedding" render process which is the direct parent. I was actually thinking of forking this to make it work but would love to put our heads together on it. This is so useful. 

I haven't done this before, should we just continue the discussion here?  I can post a bit more detail a little later if so.

**Comment by rkpatel33** [comment]
So I did a bit of digging into your code. It's a bit beyond me as in certain respects as I am more of a python guy, but here is the high level issue:

To give a little color, I have a use case where I am trying to plug into a third party messaging web application to handle some automation tasks based on my internal user needs. So I have a single browser window that has a single render process that has three <webview> tags in it, as in the code image above (1 third party site and 2 internal web apps that will eventually be true Electron object down the road). The "host" render process is the coordinator of activity and relaying messages between the `<webview>` render processes. It's def a hacky situation but kind of ideally dealt with in an Electron context. I had it set up as a Chrome extension before. Something like this:

```
- Main process
    - Render process (host)
        - webview 1 - third party website (guest)
        - webview 2 - internal (guest)
        - webview 3 - internal (guest)
```

All communication using `ipcRenderer` and `ipcMain` (as I understand) only travels between the main process and a render process (either normal render process or a render process for a `<webview>`). A `<webview>` process is considered just another render process ("guest" process) and therefore it can only send messages directly to the main process using `ipcRenderer.send`. 

If the `<webview>` process wants to send a message directly to the host/embedding render process that created it, it was to use the call `ipcRenderer.sendToHost`. It sill uses the `ipcRenderer.on` to listen for inbound messages.

Conversely, if the host render wants to talk to the guest render process, it has to use:

```
var webview = document.getElementById('webview')
webview.send('ping', {command: 'test'})
```

And for host to listen to guest, it must listen with:

```
webview.addEventListener('ipc-message', ...)
```

A hack would be to use the main process as a relay to send a message from webview > main > render, and then back again, but not ideal. Any ideas on a fix? I seems like it would be a small adjustment to your code but its not clear to me how to do it.

**Comment by harryparkdotio** [comment]
In terms of the handshake process, the whole thing is currently a bit of a mess. Disregard the handshake failure messages right now as they aren't directly connected to the messaging process. Honestly, I'm thinking about removing them as they're causing more trouble than it seems to be worth. The only use case that would see benefit would be in debugging, but even then, it may not be worth the effort required to make it work nicely.

`ipc-flux` is still in active development, with the handshake process my top priority right now, followed by `state` and `commit` functionality.

I'm still investigate webview, mostly just trying to wrap my head around it completely, especially in relation to `ipc-flux`.

I haven't had the opportunity just yet to have a proper stab at a solution, but hopefully I will be able to soon.

As far as the whole of `ipc-flux` goes, is there anything you think needs to be improved upon/reworked or renamed (specifically function handlers)?

**Comment by harryparkdotio** [comment]
**UPDATE**

I pushed a new release yesterday ***`0.3.10`*** which should fix any handshake issues.

**Comment by rkpatel33** [comment]
Great, thanks for letting me know.


_________________________
Rishi Patel | mail@rishinyc.com

**Comment by harryparkdotio** [comment]
**UPDATE**

I pushed a new release ***`0.3.13`*** with minor fixes (due to adding eslint into the testing process). I also added documentation on how to setup ipc-flux in `<webview>`, and communicating from main to webview.

**Comment by rkpatel33** [comment]
Awesome, I will check it out tomorrow!


_________________________
Rishi Patel | mail@rishinyc.com

**Comment by rkpatel33** [comment]
Took a quick look. Question -- will it still work if it is setup in the `preload` file instead of the src file? I can't remember if I tried it this way as it was early on in my reading that I tried this library. 

I have to have nodeIntegration turned off bc I am loading another website in the webview, but can still  use node in the preload file and have it persist, which is what I do. I will test with a dummy app tomorrow:

https://electron.atom.io/docs/api/webview-tag/#preload
